fog/test-client/src/test_client.rs

```diff
@@ -781,3 +864,125 @@ impl Drop for ReceiveTxWorker {
+/// in case of errors.
+/// This is thread-safe so that we can share it with the receive worker
+#[derive(Default, Debug)]
+pub struct TxInfo {
+    /// The Tx which was submitted
+    tx: Mutex<Option<Tx>>,
+    /// The block cloud returned by propose_tx
+    tx_propose_block_count: Mutex<Option<u64>>,
+    /// The block in which the tx appeared
+    tx_appeared: Mutex<Option<BlockIndex>>,
+}
+
+impl TxInfo {
+    /// Clear the TxInfo
+    pub fn clear(&self) {
+        *self.tx.lock().unwrap() = None;
+        *self.tx_propose_block_count.lock().unwrap() = None;
+        *self.tx_appeared.lock().unwrap() = None;
+    }
+
+    /// Set the Tx that we are sending (immediately after it is built)
+    pub fn set_tx(&self, tx: &Tx) {
+        *self.tx.lock().unwrap() = Some(tx.clone());
+    }
+
+    /// Set the block count returned by tx_propose (immediately after it is
+    /// known)
+    pub fn set_tx_propose_block_count(&self, count: u64) {
+        *self.tx_propose_block_count.lock().unwrap() = Some(count);
+    }
+
+    /// Set the index in which the tx appeared (immediately after it is known)
+    pub fn set_tx_appeared_block_index(&self, index: BlockIndex) {
+        *self.tx_appeared.lock().unwrap() = Some(index);
+    }
+}
+
+impl core::fmt::Display for TxInfo {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        self.tx_propose_block_count
+            .lock()
+            .unwrap()
+            .map(|proposed| {
+                write!(
+                    f,
+                    "Proposed at block index ~{}, ",
+                    proposed.saturating_sub(1)
+                )
+            })
+            .transpose()?;
+        self.tx_appeared
+            .lock()
+            .unwrap()
+            .map(|appeared| write!(f, "Appeared in block index {}, ", appeared))
+            .transpose()?;
+        self.tx
+            .lock()
+            .unwrap()
+            .as_ref()
+            .map(|tx| {
+                write!(
+                    f,
+                    "TxOut public keys: [{}]",
+                    HexList(tx.prefix.outputs.iter().map(|x| x.public_key.as_bytes()))
+                )
+            })
+            .transpose()?;
+        Ok(())
+    }
+}
+
+/// Object which helps tracking the healthy status
+#[derive(Default)]
+pub struct HealthyTracker {
+    // Set to i for the duration of the i'th transfer
+    counter: AtomicUsize,
+    // Whether we have observed any failure
+    have_failure: AtomicBool,
+    // The counter value during the most recent failure
+    last_failure: AtomicUsize,
+    // The number of clients
+    num_clients: usize,
+}
+
+impl HealthyTracker {
+    /// Make a new healthy tracker.
+    /// Sets IS_HEALTHY to true initially.
+    /// Takes num_clients which is the number of successful transfers before we
+    /// consider ourselves healthy again
+    pub fn new(num_clients: usize) -> Self {
+        counters::IS_HEALTHY.set(1);
```

Comment:
Is there a reason you are going with this way and not the health endpoint? If we already have a grpc server interface (which we do, for exposing the metrics), you could use the standardized grpc service: https://github.com/mobilecoinfoundation/mobilecoin/blob/master/util/grpc/src/health_service.rs

It is likely a bit more plumbing and make not suitable for what you're trying to do here, but I figured I'll point it out in case this plays nicer with whatever monitoring tool @jgreat wants to use.

Reply:
jgreat wanted to have a prometheus metric he could alert on and we discussed exactly what the behavior should be

Reply:
i think the util-grpc health service thing is meant to serve health checks to kubernetes, for purpose of like bouncing a failing node. i dont think we want to use that here, because the health check is for the fog service, not the test client itself

maybe i shouldn't have called the metric "IS_HEALTHY" because of this possible confusion?

Reply:
Health endpoint and the Prometheus metrics have two different use cases. The health check endpoint would be used by the k8s system to mark a service as failed (then kill are relaunch). The service its self can be healthy, still able to send and monitor transactions, but the individual transactions might fail.  

The prom /metrics is where I want this data so we can send to grafana and alert.

It doesn't really matter, but I would suggest "IS_HEALTHY" be changed to something like "last_polling_successful".